File: Assets/Bundles/Server/Resources/TS/Services/Match/BW/BWTeamService.lua
-- Compiled with unity-ts v2.1.0-75
local Reflect = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Reflect
local Service = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Service
local ServerSignals = require("Server/TS/ServerSignals").ServerSignals
local Game = require("Shared/TS/Game").Game
local Team = require("Shared/TS/Team/Team").Team
local ColorUtil = require("Shared/TS/Util/ColorUtil").ColorUtil
local Theme = require("Shared/TS/Util/Theme").Theme
local BWTeamService
do
	BWTeamService = setmetatable({}, {
		__tostring = function()
			return "BWTeamService"
		end,
	})
	BWTeamService.__index = BWTeamService
	function BWTeamService.new(...)
		local self = setmetatable({}, BWTeamService)
		return self:constructor(...) or self
	end
	function BWTeamService:constructor(teamService, matchService)
		self.teamService = teamService
		self.matchService = matchService
	end
	function BWTeamService:OnStart()
		local queueMeta = self.matchService:GetQueueMeta()
		for _, team in queueMeta.teams do
			local t = Team.new(team.name, team.id, team.color)
			self.teamService:RegisterTeam(t)
		end
		-- Temporary: even team distribution
		ServerSignals.PlayerJoin:connect(function(event)
			local teams = self.teamService:GetTeams()
			local smallestTeam = teams[1]
			for _, t in teams do
				-- ▼ ReadonlySet.size ▼
				local _size = 0
				for _1 in t:GetPlayers() do
					_size += 1
				end
				-- ▲ ReadonlySet.size ▲
				-- ▼ ReadonlySet.size ▼
				local _size_1 = 0
				for _1 in smallestTeam:GetPlayers() do
					_size_1 += 1
				end
				-- ▲ ReadonlySet.size ▲
				if _size < _size_1 then
					smallestTeam = t
				end
			end
			smallestTeam:AddPlayer(event.player)
			local color = ColorUtil:ColorToHex(smallestTeam.color)
			Game:BroadcastMessage("<b><color=" .. (color .. (">" .. (event.player.username .. ("</color></b> <color=" .. (ColorUtil:ColorToHex(Theme.Gray) .. ">joined the server.</color>"))))))
		end)
		ServerSignals.PlayerLeave:connect(function(event)
			local team = event.player:GetTeam()
			if team then
				local color = ColorUtil:ColorToHex(team.color)
				Game:BroadcastMessage("<b><color=" .. (color .. (">" .. (event.player.username .. ("</color></b> <color=" .. (ColorUtil:ColorToHex(Theme.Gray) .. ">left the server.</color>"))))))
			else
				Game:BroadcastMessage("<b>" .. (event.player.username .. ("</b> <color=" .. (ColorUtil:ColorToHex(Theme.Gray) .. ">left.</color>"))))
			end
		end)
	end
end
-- (Flamework) BWTeamService metadata
Reflect.defineMetadata(BWTeamService, "identifier", "Bundles/Server/Services/Match/BW/BWTeamService@BWTeamService")
Reflect.defineMetadata(BWTeamService, "flamework:parameters", { "Bundles/Server/Services/Global/Team/TeamService@TeamService", "Bundles/Server/Services/Match/MatchService@MatchService" })
Reflect.defineMetadata(BWTeamService, "flamework:implements", { "$:flamework@OnStart" })
Reflect.decorate(BWTeamService, "$:flamework@Service", Service, { {} })
return {
	BWTeamService = BWTeamService,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Server/Services/Match/BW/BWTeamService.ts
import { CoreServerSignals } from "@Easy/Core/Server/CoreServerSignals";
import { TeamService } from "@Easy/Core/Server/Services/Team/TeamService";
import { Game } from "@Easy/Core/Shared/Game";
import { Team } from "@Easy/Core/Shared/Team/Team";
import { ColorUtil } from "@Easy/Core/Shared/Util/ColorUtil";
import { Theme } from "@Easy/Core/Shared/Util/Theme";
import { Dependency, OnStart, Service } from "@easy-games/flamework-core";
import inspect from "@easy-games/unity-inspect";
import ObjectUtils from "@easy-games/unity-object-utils";
import { ServerSignals } from "Server/ServerSignals";
import { MatchTeamStatsDto } from "Shared/Match/MatchTeamStatsDto";
import { Network } from "Shared/Network";
import { MatchService } from "../MatchService";
import { BWService } from "./BWService";

@Service({})
export class BWTeamService implements OnStart {
	/** Used to place on the same team after re-joining the server */
	private userIdToTeamMap = new Map<string, Team>();
	private teamStatsMap = new Map<Team, MatchTeamStatsDto>();

	constructor(private readonly teamService: TeamService, private readonly matchService: MatchService) {}

	OnStart(): void {
		const queueMeta = this.matchService.GetQueueMeta();
		for (const team of queueMeta.teams) {
			const t = new Team(team.name, team.id, team.color);
			this.teamService.RegisterTeam(t);
			const stats: MatchTeamStatsDto = {
				id: t.id,
				bed: true,
				playersRemaining: 0,
			};
			this.teamStatsMap.set(t, stats);
			this.UpdateTeamAliveCount(t);
		}
		Network.ClientToServer.GetAllMatchTeamStats.Server.OnClientEvent((clientId) => {
			print("sending all stats to client: " + inspect(ObjectUtils.values(this.teamStatsMap)));
			Network.ServerToClient.UpdateMatchTeamStats.Server.FireClient(
				clientId,
				ObjectUtils.values(this.teamStatsMap),
			);
		});

		CoreServerSignals.PlayerChangeTeam.Connect((event) => {
			if (event.team) {
				this.UpdateTeamAliveCount(event.team);
			}
			if (event.oldTeam) {
				this.UpdateTeamAliveCount(event.oldTeam);
			}
		});

		CoreServerSignals.PlayerLeave.Connect((event) => {
			const team = event.player.GetTeam();
			if (team) {
				this.UpdateTeamAliveCount(team);
			}
		});

		ServerSignals.PlayerEliminated.Connect((event) => {
			const team = event.player.GetTeam();
			if (team) {
				this.UpdateTeamAliveCount(team);
			}
		});

		// Temporary: even team distribution
		CoreServerSignals.PlayerJoin.Connect((event) => {
			let assignedTeam: Team | undefined;

			let existing = this.userIdToTeamMap.get(event.player.userId);
			if (existing) {
				existing.AddPlayer(event.player);
				assignedTeam = existing;
			}

			if (!assignedTeam) {
				const teams = this.teamService.GetTeams();
				let smallestTeam = teams[0];

				for (const t of teams) {
					if (t.GetPlayers().size() < smallestTeam.GetPlayers().size()) {
						smallestTeam = t;
					}
				}

				smallestTeam.AddPlayer(event.player);
				assignedTeam = smallestTeam;
			}

			const color = ColorUtil.ColorToHex(assignedTeam.color);
			Game.BroadcastMessage(
				`<b><color=${color}>${event.player.username}</color></b> <color=${ColorUtil.ColorToHex(
					Theme.Gray,
				)}>joined the server.</color>`,
			);
		});

		CoreServerSignals.PlayerLeave.Connect((event) => {
			const team = event.player.GetTeam();
			if (team) {
				const color = ColorUtil.ColorToHex(team.color);
				Game.BroadcastMessage(
					`<b><color=${color}>${event.player.username}</color></b> <color=${ColorUtil.ColorToHex(
						Theme.Gray,
					)}>left the server.</color>`,
				);
			} else {
				Game.BroadcastMessage(
					`<b>${event.player.username}</b> <color=${ColorUtil.ColorToHex(Theme.Gray)}>left.</color>`,
				);
			}
		});

		CoreServerSignals.PlayerChangeTeam.Connect((event) => {
			if (event.team) {
				this.userIdToTeamMap.set(event.player.userId, event.team);
			}
		});
	}

	private UpdateTeamAliveCount(team: Team): void {
		const aliveCount = Dependency<BWService>().GetAlivePlayersOnTeam(team).size();
		this.UpdateTeamStats(team, (stats) => {
			stats.playersRemaining = aliveCount;
		});
	}

	public UpdateTeamStats(team: Team, callback: (stats: MatchTeamStatsDto) => void): void {
		const stats = this.teamStatsMap.get(team)!;
		callback(stats);
		print("firing update team stats: " + inspect(stats));
		Network.ServerToClient.UpdateMatchTeamStats.Server.FireAllClients([stats]);
	}
}
